util/protobuf_encoder.py

```diff
@@ -101,22 +101,82 @@ def from_json(cls, data):
         else:
             assert False, "Unknown RADType"
 
+@dataclass
+class StringPair:
+    left: str
+    right: str
+
+    @classmethod
+    def from_json(cls, data):
+        return StringPair(left=data['left'], right=data['right'])
+
+    def to_pb_bytes(self):
+        left_bytes = pb_field(field_number=1, tag=LENGTH_DELIMITED, value=str_to_bytes(self.left))
+        right_bytes = pb_field(field_number=2, tag=LENGTH_DELIMITED, value=str_to_bytes(self.right))
+        return concat([left_bytes, right_bytes])
+
 @dataclass
 class RADRetrieve:
     kind: RADType
     url: str
     script: bytes
+    body: bytes
+    headers: List[StringPair]
 
     @classmethod
     def from_json(cls, data):
-        return RADRetrieve(kind=RADType.from_json(data['kind']), url=data['url'], script=bytes(data['script']))
+        _body: bytes = b''
+        if 'body' in data:
+            _body = bytes(data['body'])
+        if data['kind'] == 'HTTP-POST':
```

Comment:
Probably nicer to handle this similarly to `_body` (so have an extra `headers=[]` before the if-statement) and unident the `return RADRetrieve` to have less duplication?

Reply:
agreed, 
less duplication and uniform variable declarations